src/main/scala/io/keen/client/scala/BatchWriterClient.scala

```diff
@@ -0,0 +1,208 @@
+  // FIXME: should be protected, but tests need to be updated; the EventStore
+  // should be an injectable component like HttpAdapter (though the injection
+  // pattern still needs work on that too)
+  val eventStore: EventStore = new RamEventStore
+  eventStore.maxEventsPerCollection = settings.maxEventsPerCollection
+
+  // Schedule sending of queued events.
+  protected val scheduledThreadPool: Option[ScheduledThreadPoolExecutor] = scheduleSendQueuedEvents()
+
+  /**
+   * Queue an event for batched publishing.
+   *
+   * @param collection The collection to which the event will be added.
+   * @param event The event
+   */
+  def queueEvent(collection: String, event: String): Unit = {
+    // bypass min/max intervals for testing
+    // FIXME: There are less kludgey ways to achieve testability here
+    environment match {
+      case Some("test") if Some("test").get matches "(?i)test" =>
+      case _ =>
+        require(
+          sendIntervalEvents == 0 || (sendIntervalEvents >= MinSendIntervalEvents && sendIntervalEvents <= MaxSendIntervalEvents),
+          s"Send events interval must be between $MinSendIntervalEvents and $MaxSendIntervalEvents"
+        )
+    }
+
+    eventStore.store(projectId, collection, event)
+
+    // If we've met a configured event count threshold, flush the queue.
+    if (sendIntervalEvents != 0 && eventStore.size >= sendIntervalEvents) {
+      sendQueuedEvents()
+    }
+  }
+
+  /**
+   * Schedule periodic sending of queued events, on a threadpool.
+   */
+  private def scheduleSendQueuedEvents(): Option[ScheduledThreadPoolExecutor] = {
+    // bypass min/max intervals for testing
+    environment match {
+      case Some("test") if Some("test").get matches "(?i)test" =>
+      case _ =>
+        require(
+          sendInterval.toSeconds == 0 || (sendInterval >= MinSendInterval && sendInterval <= MaxSendInterval),
+          s"Send interval must be between $MinSendInterval and $MaxSendInterval"
+        )
+    }
+
+    // send queued events every n seconds
+    sendInterval.toSeconds match {
+      case n if n <= 0 => None // TODO: document what config value of zero means
+      case _ =>
+        // use a thread pool for our scheduled threads so we can use daemon threads
+        val tp = Executors.newScheduledThreadPool(1, new ClientThreadFactory).asInstanceOf[ScheduledThreadPoolExecutor]
+
+        // schedule sending from our thread pool at a specific interval
+        tp.scheduleWithFixedDelay(new Runnable {
+          def run(): Unit = {
+            try {
+              sendQueuedEvents()
+            } catch {
+              case ex: Throwable =>
+                error("Failed to send queued events")
+                error(s"$ex")
+            }
+          }
+        }, 1, sendInterval.toMillis, TimeUnit.MILLISECONDS)
+
+        Some(tp)
+    }
+  }
+
+  /**
+   * Flush queued events, removing them from the queue as they are successfully sent.
+   */
+  def sendQueuedEvents(): Unit = {
```

Comment:
Excellent recommendation. We want to maintain parity as much as possible, but we also want to take advantage of benefits afforded by Scala too.